File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.js
function solution(sequence) {
    const n = sequence.length;
    let p1 = new Array(n); // 1, -1, 1, -1...
    let p2 = new Array(n); // -1, 1, -1, 1...
    
    // 1. 펄스 수열 곱한 배열 생성
    for (let i = 0; i < n; i++) {
        p1[i] = i % 2 === 0 ? sequence[i] * 1 : sequence[i] * -1;
        p2[i] = i % 2 === 0 ? sequence[i] * -1 : sequence[i] * 1;
    }
    
    // 2. 카데인 알고리즘 함수
    function getMaxSum(arr) {
        let maxVal = arr[0];
        let currentMax = arr[0];
        
        for (let i = 1; i < arr.length; i++) {
            currentMax = Math.max(arr[i], currentMax + arr[i]);
            maxVal = Math.max(maxVal, currentMax);
        }
        return maxVal;
    }
    
    // 3. 두 경우 중 최댓값 반환
    return Math.max(getMaxSum(p1), getMaxSum(p2));
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
class Solution {
    public long solution(int[] sequence) {
        long answer = 0;
        int n = sequence.length;
        
        // 펄스 패턴 1: 1, -1, 1, -1, ...
        // 펄스 패턴 2: -1, 1, -1, 1, ...
        long[] p1 = new long[n];
        long[] p2 = new long[n];
        
        for (int i = 0; i < n; i++) {
            if (i % 2 == 0) {
                p1[i] = (long) sequence[i] * 1;
                p2[i] = (long) sequence[i] * -1;
            } else {
                p1[i] = (long) sequence[i] * -1;
                p2[i] = (long) sequence[i] * 1;
            }
        }
        
        // 카데인 알고리즘(Kadane's Algorithm) 적용
        answer = Math.max(getMaxSum(p1), getMaxSum(p2));
        
        return answer;
    }
    
    private long getMaxSum(long[] arr) {
        long maxVal = arr[0];
        long currentSum = arr[0];
        
        for (int i = 1; i < arr.length; i++) {
            currentSum = Math.max(arr[i], currentSum + arr[i]);
            maxVal = Math.max(maxVal, currentSum);
        }
        
        return maxVal;
    }
}
